src/Processors/QueryPlan/Optimizations/calculateHashTableCacheKeys.cpp

```diff
@@ -0,0 +1,199 @@
+#include <Processors/QueryPlan/Optimizations/Optimizations.h>
+
+#include <Core/Joins.h>
+#include <IO/WriteHelpers.h>
+#include <Interpreters/SetSerialization.h>
+#include <Interpreters/TableJoin.h>
+#include <Processors/QueryPlan/ITransformingStep.h>
+#include <Processors/QueryPlan/JoinStepLogical.h>
+#include <Processors/QueryPlan/ReadFromRemote.h>
+#include <Processors/QueryPlan/Serialization.h>
+#include <Processors/QueryPlan/SourceStepWithFilter.h>
+#include <Storages/IStorage.h>
+#include <Common/SipHash.h>
+
+using namespace DB;
+
+namespace
+{
+
+UInt64 calculateHashFromStep(const ReadFromParallelRemoteReplicasStep & source)
+{
+    SipHash hash;
+    hash.update(source.getSerializationName());
+    hash.update(source.getStorageID().getFullTableName());
+    return hash.get64();
+}
+
+UInt64 calculateHashFromStep(const SourceStepWithFilter & read)
+{
+    WriteBufferFromOwnString wbuf;
+    SerializedSetsRegistry registry;
+    IQueryPlanStep::Serialization ctx{.out = wbuf, .registry = registry};
+
+    writeStringBinary(read.getSerializationName(), wbuf);
+    if (const auto & snapshot = read.getStorageSnapshot())
+        writeStringBinary(snapshot->storage.getStorageID().getFullTableName(), wbuf);
+    if (const auto & dag = read.getPrewhereInfo())
+        dag->prewhere_actions.serialize(ctx.out, ctx.registry);
```

Comment:
Maybe we can actually calcualte hash from ActionsDAG instead of serialization.
E.g. something like this
https://github.com/ClickHouse/ClickHouse/pull/69236/files#diff-fdf54241f19355cdaf1bd2ac825352747f9ff9d1bcbff7a677f8e573eae17b18R139